### ISSUE TITLE
use memory-only tmpfiles to track irodsAgent PIDs

### ISSUE THREAD

**Issue by jasoncoposky** [irods/irods#4080]
- [ ] master
- [ ] 4-2-stable
-----

The native rule engine plugin makes a copy of the rule base, which can occasionally collide with a previous copy should one exist.  If the NREP is configured the irodsctl script should clean up any copies of the rule bases in /etc/irods on server stop.

**Comment by tempoz** [comment]
Simply deleting the files when the server is started or stopped is not sufficient to ensure collisions are avoided, as pids can and do wrap during a single execution of the server. We should make a dedicated directory for cached rulebases and use tmpfile to create the actual cache files to avoid collisions.

**Comment by trel** [comment]
Consider:
`/etc/irods/cached_rulebases/core.re.<tmpfilevalue>`

And clear them out on server startup (not stop).

**Comment by tsmeele** [comment]
Collisions indeed occur after Linux process numbers have wrapped:
```
Apr  6 05:59:45 pid:8674 NOTICE: writeLine: inString = Agent process started for puser=rods and cuser=rods from 127.0.0.1

terminating with uncaught exception of type boost::filesystem::filesystem_error: boost::filesystem::copy_file: File exists: "/etc/irods/rules-dgk.re", "/etc/irods/rules-dgk.re.8683"
Caught signal [6]. Dumping stacktrace and exiting
```
`

**Comment by trel** [comment]
partial solution (mentioned above) from two years ago:
  https://github.com/tempoz/irods/commits/cache_fix

**Comment by trel** [comment]
partial solution with #5224 is landed for 4.2.9 - regex removal of any leftover pidfiles on server startup.

we'll pursue a more complete solution with only tempfiles (no actual pidfiles at all) to close this issue.

**Comment by kript** [comment]
As discussed elsewhere, it's not unusual for us to have services remaining running for months, even years, so being unique is a good thing longer term.

**Comment by korydraughn** [comment]
This issue was resolved by #5210.

Rulebase files are stored in memory now.
